### PR TITLE
[Networking] - useAPIClients Init Custom Hook + Auth changes to enable mutations/subscriptions

### PR DESCRIPTION
**Pull request by drewjhart**
This completes the work on the `Auth` layer. For a full description of this layer, as well as an overview of the new API client and a step-by-step review of the setup process, see the below doc in the Wiki:

[https://github.com/rightoneducation/righton-app/wiki/Backend-%E2%80%90-Authentication-and-Authorization](https://github.com/rightoneducation/righton-app/wiki/Backend-%E2%80%90-Authentication-and-Authorization)

The PRs break down to:
-> [[Backend] - Google OAuth Integration #988](https://github.com/rightoneducation/righton-app/pull/988)
->-> [[Central_v2] - Container for authorization #1013](https://github.com/rightoneducation/righton-app/pull/1013)
->->-> [[Authentication/Authorization] - Networking updates #1058](https://github.com/rightoneducation/righton-app/pull/1058)
->->->->[ [Networking] - Upgrade Amplify v5 to v6 #1059](https://github.com/rightoneducation/righton-app/pull/1059)
->->->->-> [[Networking] - AuthAPIClient #1060](https://github.com/rightoneducation/righton-app/pull/1060)
->->->->->-> [[Networking] - useAPIClients Init Custom Hook + Auth changes to enable mutations/subscriptions](https://github.com/rightoneducation/righton-app/pull/1068)

**Review Cheat Sheet**
Hey @maniramezan , this set of PR's is a bit confusing for two reasons:

1. A lot of the file changes you are seeing will be auto-generated from Amplify (resolvers and config files etc)
2. A lot of the previous PRs have initial implementations of Auth methods that are refined in later PRs

Therefore:

1. I'd focus primarily on this PR as the main PR to review. This represents the final AuthAPIClient and the initialization and deployment of it throughout the code (using the custom hook). Implementations in previous PRs are overwritten in this one. 
2. I've added a bunch of comments throughout this PR to explain how this works and the set up of it.
3. If you want insight into the actual auth setup, I've created a Wiki document here: [https://github.com/rightoneducation/righton-app/wiki/Backend-%E2%80%90-Authentication-and-Authorization](https://github.com/rightoneducation/righton-app/wiki/Backend-%E2%80%90-Authentication-and-Authorization)

Hopefully, this makes your job a bit easier when reviewing this stuff. Definitely let me know if anything is confusing though!

### CODE REVIEW

File: central/src/App.tsx [hunk 1]
```diff
@@ -27,18 +27,20 @@ const theme = createTheme({
 
 function App() {
   const [alert, setAlert] = useState<Alert | null>(null);
-  const apiClients = new APIClients(Environment.Developing);
   const alertContext = {
     alert,
     setAlert,
   };
```
Comment (drewjhart) [inline central/src/App.tsx:34]: useApiClient custom hook for initialization of AuthAPIClient first, and then other APIClients

File: central_v2/src/pages/Login.tsx [hunk 2]
```diff
@@ -4,7 +4,7 @@ import { styled } from "@mui/system";
 import { Link } from "react-router-dom";
```
Comment (drewjhart) [inline central_v2/src/pages/Login.tsx:4]: central_v2 will be reworked in a subsequent set of PRs in Q3 (Design is still working on it)

File: networking/amplify/backend/api/mobile/custom-roles.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (drewjhart) [inline networking/amplify/backend/api/mobile/custom-roles.json:1]: this role is required to ensure that the `createGame` Lambda function is explicitly provided with its Execution Role

File: networking/amplify/backend/api/mobile/resolvers/Mutation.createTeam.auth.1.req.vtl [hunk 4]
```diff
@@ -7,12 +7,12 @@ $util.qr($ctx.stash.put("hasAuth", true))
 $util.unauthorized()
 #end
```
Comment (drewjhart) [inline networking/amplify/backend/api/mobile/resolvers/Mutation.createTeam.auth.1.req.vtl:8]: I think there is a bug with Amplify that requires us to manually set these resolvers. Pretty annoying so I'll talk to our AWS contact next time. See Wiki for a link to an Amplify issue that confirms this and recommends the below edits to resolve

File: networking/amplify/backend/api/mobile/schema.graphql [hunk 5]
```diff
@@ -1,10 +1,18 @@
 type Mutation {
-  createGameSessionFromTemplate(input: CreateGameSessionFromTemplateInput!): String @function(name: "createGame-${env}")
+  createGameSessionFromTemplate(input: CreateGameSessionFromTemplateInput!): String
```
Comment (drewjhart) [inline networking/amplify/backend/api/mobile/schema.graphql:2]: @auth directive required for mutations and queries as well as the @types

File: networking/src/APIClients/BaseAPIClient.ts [hunk 6]
```diff
@@ -44,11 +45,13 @@ export interface IQueryParameters {
```
Comment (drewjhart) [inline networking/src/APIClients/BaseAPIClient.ts:45]: pass AuthAPIClient to other APIClients to provide auth to the API requests

File: networking/src/Hooks/useAPIClients.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
+import { useState, useEffect } from 'react';
```
Comment (drewjhart) [inline networking/src/Hooks/useAPIClients.ts:1]: custom Hook to provide proper sequence for APIClient init. First init AuthAPIClients and then pass those clients down into the other APIClients

File: networking/src/APIClients/gamesession/TeamMemberAPIClient.ts [hunk 8]
```diff
@@ -40,7 +40,7 @@ export class TeamMemberAPIClient extends BaseAPIClient {
   subscribeUpdateTeamMember(
     id: string,
     callback: (result: ITeamMember) => void
-  ): Promise<any> {
+  ) {
```
Comment (maniramezan) [inline networking/src/APIClients/gamesession/TeamMemberAPIClient.ts:43]: Intentionally removing the return type?

File: networking/src/APIClients/auth/interfaces/IAuthAPIClient.ts [hunk 9]
```diff
@@ -1,6 +1,7 @@
 
 export interface IAuthAPIClient {
   isUserAuth: boolean;
+  init(): Promise<void>;
```
Comment (maniramezan) [inline networking/src/APIClients/auth/interfaces/IAuthAPIClient.ts:4]: See the comment above regarding combining this with constructor into one method

**Review by maniramezan — APPROVED** [review]
Great job on this, this looks great, and I love how you have organized the files and methods ❤️

**Pull request closed without merging**